### DISCUSSION TITLE
Better Hypercomplex Numbers

### DISCUSSION THREAD

**Discussion by JoshZeidner** [sympy/sympy#22760] [Ideas]
Hello, 

I'm fairly new to Sympy. I've been working with it for a few weeks. I'd like to make a few points about Sympy, Hypercomplex numbers and the Sympy framework. I'm not exactly a critic, I think the work with Quaternions is a useful package. Seems as though @oscargus is the primary developer for the quaternion package so hats off to him for the work done so far.

# Better Quaternions

Quaternions are a 4 dimensional extension of the Complex Numbers.

Currently Sympy supports Quaternions in much the same way Mathematica does. They are given a special category and set of functions that works with other aspects of sympy in a limited way. These functions are similar to many other comparable Quaternion libraries in other languages.

Quaternions are really an extension of standard algebra in much the same way Complex Numbers are. While the 3 imaginary components of a Quaternion do not exist on the R (Real) number line, they are related to standard algebra and can even be expressed as a real number in some cases.

**So my suggestion: a better way to support quaternions is to make 2 more **ImaginaryUnit** classes.** Thus instead of simply I, we also have two other Imaginary Units J and K. We must also implement their ***non-commutative*** properties. These are:

I * J = K

J * I = -K

J * K = I

K * J = -I

K * I = J

I * K = -J

other related identities hold eg. K / J = I and K / I = -J (an extension of the first multiplication rule).

If these properties were implemented, then we gain access to all the other features of Sympy. Remarkably all of the rotation and scaling operations we are familiar with from Complex Analysis also work for the Quaternions through these simple algebraic rules. Besides these unique non-commutative properties, Quaternions behave otherwise like the imaginary unit I. The field of Quaternionic Analysis is not well explored at this point.

For example in the Quaternion documentation it mentions support for Quaternions over Complex Fields.

```
>> q2 = Quaternion(3 + 4*I, 2 + 5*I, 0, 7 + 8*I, real_field = False)

>> q1
x + x**3*i + x*j + x**2*k

>> q2
(3 + 4*I) + (2 + 5*I)*i + 0*j + (7 + 8*I)*k
```

I'm not completely sure what is meant by "Quaternions over Complex Fields". The Complex Field C is a subset of Q. Thus the final expression is in reality a normal Quaternion: 

```
(3 + 4*I) + (2 + 5*I)*i + 0*j + (7 + 8*I)*k == ( -2 +6i -8j +7k )
```

**_And this highlights the basic problem, Sympy does not recognize that the "I" in the Quaternions is the same as the I in Complex numbers._** They are the same character in the story.

There was something Hamilton himself explored called a Biquaternion, where there are two separate I units (naming one of them h). They have limited use (and are of limited interest). To be quite frank, there are many (even credible) works done in the field of hypercomplex numbers where the big picture is lacking. Cayley himself often expressed Quaternions using only two imaginary units and a pair of complex numbers ie. (5+i) + j(6+2i) = (5+i+6j+2k)

# Hypercomplex Numbers

Most people interested in quaternions are 3d graphics programmers and they are only interested in a small subset of the functions of Quaternions. Sympy targets a broader crowd. With this methodology, we can easily extend to the Octonions as well (would require 7 imaginary units, 3 of which can be used as is from Quaternions and Complex numbers). This doubling pattern, known as the Cayley-Dickson Construction continues ad infinitum and one can continue the process indefinitely simply by expanding the number of imaginary units. Note that the Octonions are the final division algebra. and are **non-associative**. The numbers that exist beyond Octonions pose interesting challenges to our normal concept of mathematics.

I may have some time to make some progress with this on my own, however it is far more effective effort for someone with knowledge of the internals. If there is enough interest in this concept, it's best that a developer with experience in Sympy make the major cuts. I am very happy to be part of the project as a developer as well.

Please let me know if any of the core project members are interested in supporting this effort. Most (all?) of the existing function signatures can be retained. It will allow for much simpler programmatic expressions IMHO. We can also offer better coupling to the Lie group functionality.

-jmz

**Comment by oscargus** [comment]
First, not sure who has done most work here, but not me. I have at the most cleaned up some code.

If I get it correctly, what you are interested in is basically introducing `J` and `K` to act similar to `I`? That may indeed make sense. 

The main problems I can see are:
* right now there are functions like `im` that get the imaginary part, what are the corresponding for J and K? (Should `im(K)` return `J`?)
* the assumption system has the concept of complex and imaginary,  so symbols can be assumed to be those (or not). Adding hypercomplex to both assumption systems may be quite demanding. Or not.

An option may possibly be to not have this on by default, but one can import J and K explicitly and those have mul, add etc defined.

Just some initial thoughts.

**Reply by JoshZeidner** [reply]
> If I get it correctly, what you are interested in is basically introducing J and K to act similar to I? That may indeed make sense.

that is the suggestion. This will likely segue into an implementation for Octonions.

> right now there are functions like im that get the imaginary part, what are the corresponding for J and K? (Should im(K) return J?)

right. I would think that a Complex Number should be structured underneath like any other term with one addition operation ie. ( 1 + x ) should really be treated exactly like ( 1 + i  ) in the algebra system. There is a tendency to treat complex and hypercomplex numbers as though they are an array of numbers eg. ( 0 + 2i + 0j + 1k ) when in fact they behave like any other term of that kind. For instance, ( 45 + 12i  ) behaves exactly like the Quaternion ( 45 + 12i + 0j + 0k ). The non-commutative properties only emerge when I, J and K interact. in fact one can use ( 45 + 12j ) and it will act just like complex numbers as long as the other two Imaginary Units do not appear in any calculations.

There are other interesting considerations as well. For instance -1 has infinite square roots if you consider the field of Quaternions. There are two if we consider the Complex field ( the roots are i and -i ). ( btw- Sympy only returns one root which is technically inaccurate ). Probably Sympy should return a conditional set in this case.

> An option may possibly be to not have this on by default, but one can import J and K explicitly and those have mul, add etc defined.

this is exactly how I think it should work. If J and K are in the namespace assume the application is interested in a Quaternion answer, if not assume the current Complex behavior.

We're really on the edge of Mathematics here which requires extra effort, is challenging, but may bring Sympy into the front of the class for these sorts of applications. These were roughly issues Riemann himself dealt with. It says something interesting about hypercomplex analysis, there is assumptions there about what the Mathematician is looking for.

**Reply by oscarbenjamin** [reply]
I think that it would be good to extend the `algebras` module to handle the Cayley-Dickinson construction or to make special units `I, J, K` there so that those could be used instead of the current `Quaternion` class. There could also be a "kind" related to these so that they can be used more broadly but in a well defined way. Currently the kind of a `Quaternion` is undefined:
```python
In [13]: Quaternion(0).kind
Out[13]: UndefinedKind
```
I don't think though that it is a good idea to try and seamlessly integrate quaternions and complex numbers. That would complicate all of the internals of everything. Every function would need a lot of checking to figure out if its arguments are ordinary complex numbers or not. It isn't even clear how to define all functions if they need to accept quaternion arguments. Many rewrite operations would no longer be valid and it would become possible to make all kinds of invalid expressions. Things like the whole of the `evalf` stack would need to be rewritten since it currently is formulated around complex numbers.

Extending the support for quaternions and related algebras would be good but conflating quaternions with the ordinary complex numbers that the vast majority of SymPy code is organised around seems like a bad idea that would cause many problems for very little benefit. Adding a "kind" for quaternions or algebras so that they can be easily distinguished from other expressions would make it possible to extend the use of quaternions in a well-defined way without breaking everything else.

**Reply by JoshZeidner** [reply]
I hear you. I did take a look at the code and even just a cursory search on `ImaginaryUnit` reveals that the behavior for Complex Numbers is strewn throughout the entire code base. I thought perhaps it could be subclassed giving us a J and a K unit. I think you're suggesting that this would be difficult.


The overall behavior of Quaternions package doesn't seem to be 100% working:

https://gist.github.com/JoshZeidner/c68cf84fabb1fb79228b696e88ba1e10

this one shows a few problems some of them are terminal errors

https://gist.github.com/JoshZeidner/c44ab96702dac9f0bc3eb26fdfa61e7e


> I don't think though that it is a good idea to try and seamlessly integrate quaternions and complex numbers. That would complicate all of the internals of everything. Every function would need a lot of checking to figure out if its arguments are ordinary complex numbers or not. It isn't even clear how to define all functions if they need to accept quaternion arguments.

Not sure we're entirely on the same page here. If I, J, K were just specialized symbols there would be no "Quaternion" class at all. That's roughly the meaning of my original post.  A Quaternion is merely an algebraic term that uses at least two of the Imaginary Units I, J, K. As for support for algebraic functions for hypercomplex numbers, they are defined but differ quite a bit from the Reals (eg. use Taylor Series). Differentiation requires special treatment due to non-commutativity.

> There could also be a "kind" related to these so that they can be used more broadly but in a well defined way

Don't feel like I understand your suggestion here.

The advantage of course is reusability, modularity etc. As I mentioned above we could easily expand to Octonions with this method. Granted, I don't have a firm grasp of what would be involved code wise. Seems like your opinion is this is not worth it. Maybe this idea will appear in the next major code refactor. In general it seems like the class structures need some work.

**Reply by oscarbenjamin** [reply]
Although I, J and K could just be special symbols they need to be clearly distinct from ordinary complex numbers. As you say special rules are needed for differentiation etc which means that there is a lot of code that will need to check if it is dealing with quaternions or not. So if I have something like `Add(Integer(1), Mul(2, J))` how do I know if this is a quaternion?

My suggestion is that the `.kind` attribute can be used to clearly distinguish between quaternions (or some more general framework that includes quaternions) and ordinary numbers. We currently have:
```python
In [1]: (1 + 2*I).kind
Out[1]: NumberKind

In [2]: Matrix([[0]]).kind
Out[2]: MatrixKind(NumberKind)
```
We could add a `kind` for quaternions that uses the same system for determining if something is a quaternion or not although there would still be ambiguity in the case of real quaternions.

Then code that needs to know if it is dealing with quaternions can check `.kind`. We need quaternions to *not* be `NumberKind` though. The definition of `NumberKind` is basically the complex numbers and the standard infinities because that is a relatively well-defined scope that covers most of what SymPy needs to do. Including quaternions here would make `NumberKind` not necessarily commutative and would complicate core logic for things like differentation. The idea of `NumberKind` is that we can check for it to rule out weird cases like quaternions.

The `kind` system is new and is not fully implemented but the intention is that it should be able to handle precisely this sort of thing by making it easy to tell apart objects that should be handled differently.

**Reply by oscarbenjamin** [reply]
> The overall behavior of Quaternions package doesn't seem to be 100% working:
> 
> https://gist.github.com/JoshZeidner/c68cf84fabb1fb79228b696e88ba1e10

That's just a printing problem:
```python
In [6]: q
Out[6]: 0 + 0*i + 1*j + 0*k

In [7]: sqrt(q)
Out[7]: 
                   1/2
0 + 0*i + 1*j + 0*k   

In [8]: type(sqrt(q))
Out[8]: sympy.core.power.Pow

In [9]: sqrt(q).args
Out[9]: (0 + 0*i + 1*j + 0*k, 1/2)
```
The `sqrt` returns a symbolic power. The printing code should put brackets around the quaternion.

> this one shows a few problems some of them are terminal errors
> 
> https://gist.github.com/JoshZeidner/c44ab96702dac9f0bc3eb26fdfa61e7e

There are some misunderstandings here like you've used `2/3` which in Python makes a float. You need to use `Rational(2, 3)` or `S(2)/3` or something.

You've messed up the arguments to `ikRotate`:
```python
In [22]: ikRotate = Quaternion( ( 0, ( 3**(1/2) )/2), 0, 1/2 )

In [23]: ikRotate
Out[23]: (0, 0.866025403784439) + 0*i + 0.5*j + 0*k
```
That should just give an error because you've passed a tuple as the first argument (the Quaternion constructor should be checking for `NumberKind`).

**Reply by JoshZeidner** [reply]
> So if I have something like Add(Integer(1), Mul(2, J)) how do I know if this is a quaternion?

I'm assuming this means ( 1 + 2 * J )? The term itself behaves exactly like a complex number (in fact Electrical Engineers use J not I to denote Imaginary Unit). So as I mentioned above, you don't really need to handle any Quaternion specific functionality unless either I or K is added somehow to the expression. This could, I assume, be implemented through a common imaginary number subclass.

It's like this. If you have only one imaginary unit, you are dealing with a PLANE. When you add just one more, you are now dealing with 4 dimensional space (3 dimensions are not really algebraically perfect but can be simulated using Quaternions). There exist infinite Complex Planes inside of the field of Quaternions. If you only operate on one, in your case R/J plane then you are really doing Complex Algebra. It's related to concept of `Vector Span`. The system could presumably detect the presence of more than one Imaginary Unit and do the appropriate actions. @oscargus seemed to understand this idea well in his above post.

![image](https://user-images.githubusercontent.com/3150100/147590782-ff3adf8d-c7a5-4274-a34d-679372124504.png)

https://www.youtube.com/watch?v=k7RM-ot2NWY

> Then code that needs to know if it is dealing with quaternions can check .kind. We need quaternions to not be NumberKind though. (...)

Your points about the .kind feature seem a bit over my head... can't really add much to the discussion re. this.

**Reply by JoshZeidner** [reply]
It's important to understand that Quaternions are not a matrix of 4 numbers. They are **additional dimensions** of Real algebra. It's not uncommon to have this sort of limited perspective on them.

Thanks for having a look at those issues. Yes, Python admitted an erroneous parenthesis.

**Reply by oscarbenjamin** [reply]
> So as I mentioned above, you don't really need to handle any Quaternion specific functionality unless either I or K is added somehow to the expression.

There needs to be a quick and unambiguous way of checking that something is *not* a quaternion. If we want to make quaternions work without a Quaternion class then we need some way to distinguish them from ordinary numbers. The fact that there are subspaces of the quaternions that are isomorphic to the complex numbers is not really usable as a way of operating with (and without) quaternions inside functions like `exp`, `sqrt` etc.

**Reply by JoshZeidner** [reply]
>  The fact that there are subspaces of the quaternions that are isomorphic to the complex numbers

but that offers you a unique opportunity! because you can factor out the non-commutativity algebraically and lose the constraints of non-commutativity. This is only one advantage to this sort of system I am describing. This benefit is even more valuable in the field of Octonions.

**Reply by JoshZeidner** [reply]
> not really usable as a way of operating with (and without) quaternions inside functions like exp, sqrt etc.

As I recall you've got the classes `Atom` and `Expr`. You're saying there is no way to detect the presence of more than one ImaginaryUnit? There must be some kind of tree structure in there that stores the expression.

**Comment by JoshZeidner** [comment]
a few others who have contributed: @praneethratna (thanks for the help on the bug report!), @NikhilPappu @Geektrovert @ArighnaIITG @oscarbenjamin @supreet11agrawal @ShubhamKJha @maxencemayrand @GaurangTandon @smitgajjar @smichr @RohitJain1103 @Upabjojr @ylemkimon @eric-wieser @sylee957 @Shikhar1998 @nickovs @mohitshah3111999 @ethankward @dhyams @bmgxyz @2torus

**Comment by sylee957** [comment]
I think that there should be conflicting ideas to represent Quaternions inside 4-vector or some multiplications with monomials `I, J, K`, or even other representations like polar representations. But what is the actual advantage of picking `I, J, K` over 4-vector representation or vice versa? 
If we consider the hypercomplex numbers to appear, we can't even implement every alphabets `I, J, K, L, ...`

**Reply by JoshZeidner** [reply]
Hi, it seems to be a simpler more powerful way to manage Quaternions and other hypercomplex numbers. This of course depends on the underlying structure of Sympy. @oscargus seems to think it's a good idea, @oscarbenjamin seems to be critical.

> If we consider the hypercomplex numbers to appear, we can't even implement every alphabets I, J, K, L, ...

I imagine we'll need to use an alternative notation. Perhaps for the Octonions we'll use `O[0]`, `O[1]` ... `O[7]`. The next number system in the Cayley-Dickson construction has 15 imaginary units.

**Comment by JoshZeidner** [comment]
just a sample of what I had in mind for the syntax:



```
from sympy.algebras import I, J, K

#create a 4 dimensional number
q = 1+2*J+4*K

#scale it  ....
q = q*6

#rotate 90 degrees ...
q = q*K 

#take a complex number and move it into 4 dimensions
p = ( I ** Rational(2,3) ) * J

# p is a 4 dimensional number, now move it back to 2d complex plane, regains commutativity
p = p / J
```

**Reply by oscarbenjamin** [reply]
What you show here all seems reasonable with the proviso that `I` here should not be the same as `sympy.I` the unit for the ordinary complex numbers in sympy. Note that in SymPy objects are commutative or not and if commutative they commute with everything including things that are not commutative:
```python
In [5]: I.is_commutative
Out[5]: True

In [6]: J, K = symbols('J, K', commutative=False)

In [7]: J*K*I
Out[7]: ⅈ⋅J⋅K
```
It is not easy or desirable to change `I` to be non-commutative. This behaviour is baked into the `Mul` class which should not be further complicated by needing to check for more special cases like quaternions. Instead `sympy.algebras.I` should be a different non-commutative unit.

**Reply by JoshZeidner** [reply]
>  Note that in SymPy objects are commutative or not and if commutative they commute with everything including things that are not commutative

that seems ok to me. Note that when reals operate on Quaternions they are also commutative.

```
q = Quaternion( 2,3,4,7 )

6 * q == q * 6  #true
```

I appreciate all the information you're providing about the internals. I'm not sure if we need to change I to be non-commutative. Only if we multiply it with J or K. Therefore the non-commutative property really only lives inside J and K. This is also true about Octonions and Associativity. 
So if we look at the Quaternions module we have `def _generic_mul(q1, q2):` which is the Quaternion multiplication. I presume the q1 param is the LEFT operand and the q2 is the RIGHT operand. If we made a class for J and K it would have a similar function. Does the system arbitrarily put the operands into either q1 or q2? there's no guarantee to the order here?

Lets say were writing a J class(just a sketch not working code). so 

```
def _generic_mul(q1, q2):
  if isinstance(q1, J) && isinstance(q2, J) return -1
  if isinstance(q1, K) && isinstance(q2, J) return -I
  if isinstance(q1, J) && isinstance(q2, K) return I
  if isinstance(q1, I) && isinstance(q2, K) return -J

  ... and so on
```

> further complicated by needing to check for more special cases like quaternions.

so this would not require a change to the `ImaginaryUnit `class.

I imagine implementing non-associativity would be harder. I couldn't even locate in the code where you could do that.

**Reply by JoshZeidner** [reply]
I guess the issue might be the algebraic simplification part of the system. The system must know that J*K != K*J and therefore doesn't have the liberty to move that around when it's attempting to simplify an equation. To my point below it seems Sympy doesn't have that powerful of a class structure. There is no way to define inside of Quaternions (tmk) the rules the system can use to simplify a term using that type. This stuff is basic Compiler Theory. I did see Antlr in there.

There seems to be the two categories of action in Sympy. It **calculates** and it **simplifies**.  Have a look at this:

https://gist.github.com/JoshZeidner/dc86d368088e508920733e3a39850fa6

Impressively, Sympy accepts variables inside Quaternions (this behavior seems to be automatic not really directly supported in the Quaternion class). So, `q = Quaternion(A,4* A, 3* A, 1)` is actually a Line in 4d space. The system does not simplify the operation properly.

**Reply by oscarbenjamin** [reply]
If the `Quaternion` class is not used any more then something like `1 + 2*I + 3*J` would be represented using `Add` and `Mul`. That's how it works with complex numbers:
```python
In [20]: srepr(1 + 2*I)
Out[20]: 'Add(Integer(1), Mul(Integer(2), I))'
```
See here:
https://docs.sympy.org/latest/tutorial/manipulation.html

The logic for handling the arguments of a `Mul` is mostly in `Mul.flatten`:
https://github.com/sympy/sympy/blob/27613158382a668605a21616706a8d64a3573997/sympy/core/mul.py#L192-L194
That function already has too many special cases. It should not be made any more complicated to handle this.

**Reply by JoshZeidner** [reply]
> That function already has too many special cases.

Yes, I already took a look at `Mul.flatten`. Generally I think the whole structure needs some reworking. There needs to be a more generalized class structure. I can't even find where the logic for complex multiplication is.

**Reply by oscarbenjamin** [reply]
Complex multiplication just makes a `Mul`:
```python
In [29]: e = (1 + 2*I)*(3 + 4*I)

In [30]: e
Out[30]: (1 + 2⋅ⅈ)⋅(3 + 4⋅ⅈ)
```
The `expand` function can be used to multiply out the brackets (flatten a Mul of Adds):
```python
In [31]: expand(e)
Out[31]: -5 + 10⋅ⅈ
```
Internally `expand` computes all the cross-products and then essentially evaluates to this:
```python
Add(Mul(1, 3), Mul(1, 4*I), Mul(2*I, 3), Mul(2*I, 4*I))
```
Here `Mul.flatten` will collapse all rationals so `Mul(1, 3)` just becomes 3 (i.e. `Integer(3)`). Then `Mul.flatten` also collects common factors into a `Pow` so `Mul(2*I, 4*I)` becomes `Mul(8, Pow(I, 2))`. Then the `I._eval_power` method determines that `I**2 = -1`:
https://github.com/sympy/sympy/blob/27613158382a668605a21616706a8d64a3573997/sympy/core/numbers.py#L4434
So the `Add` gets arguments like
```python
Add(3, Mul(4, I), Mul(6, I), -8)
```
Finally `Add.flatten` collects together all rationals (or float etc) and all terms that differ only by rational coefficients to produce:
```python
Add(-5, Mul(10, I))
```
https://github.com/sympy/sympy/blob/27613158382a668605a21616706a8d64a3573997/sympy/core/add.py#L179

In principle the idea is that different subclasses of `Expr` can implement their own behaviour by defining special methods like `_eval_power`. In practice lots of special cases still need to be handled in the core routines (like `Add.flatten`) which isn't scalable as more and more types of expression are added. Also the problem is that any special methods would only ever be defined for the type of the head of the expression.

The code in `Add.flatten` etc predates the "kind" system which is incomplete but is intended to give a more principled way of extending the core from outside: the core should handle `NumberKind` and when a routine like `Add.flatten` encounters other kinds then the Kind object itself should be consulted to determine how to process the args. That way the `Kind` can be used to determine if an object should be handled with non-commutative or non-associative manipulations. This way if we have `Add(Integral(Matrix(...)), ...)` then `Add.flatten` can see that the kind of the object is MatrixKind and MatrixKind can determine how to handle this e.g. by checking that all args have the same shape. This isn't really possible with special methods because e.g. in this example if Add calls special methods on its args then the special methods of Integral will be called rather than those of Matrix.

**Reply by JoshZeidner** [reply]
@oscarbenjamin, thanks a lot for all your detailed responses. I'll take a closer look at your message later, seems to clear up a few things for me. I imagine that somewhere is some kind of expression tree and Sympy computes alternatives (simplifications, factoring, etc.).

I'm going to attempt to implement the basic idea using Quaternions.py as a starting point. It has a simple class hierarchy. Currently I'm wondering: how do I control the pretty print output?

Feel free to have a look, leave some comments and whatever other kinds of input you see appropriate. I just started this. Anyone else who is contributing here is also welcome.

https://colab.research.google.com/drive/1rV_lT3L_jLr5AVJ_Z_NqALtSal69_GIy?usp=sharing

**Reply by oscarbenjamin** [reply]
The easiest way to show others what you are doing and collect feedback is by opening a pull request that makes the changes to the SymPy codebase.

**Reply by JoshZeidner** [reply]
ok I will probably do that soon, I'm going to see how much progress I make here... advantage to Colab is you can run it right in the document. No environment issues.

**Comment by oscargus** [comment]
I think we should try to analyze the consequences before really making a "decision". From some perspective it is "just" to add two more "imaginary" dimensions. The problem is that almost all of SymPy's code base is formed around the concept of having only a single imaginary dimension. (One can think of a complex number as either a multinomial in 1 and I or a two-element vector, similarly a Quaternion is a multinomial in 1, I, J and K or a four-element vector.)

I think checking if something is hypercomplex is really not that hard, at least not if we restrict our self to real, complex and hypercomplex numbers (problem may happen with whatever the similarity to imaginary is called).

I also believe it makes sense to explicitly import J and K from somewhere when needed to avoid confusion with accidentally forgetting to define J and K and using them as ordinary symbols. Weird things may happen and we probably want to off-load the bug reports and forum with these things. 

I'm a bit worried what consequence this may have for all other functions in SymPy. There may be quite a bit of changing codes to just support (or error) on hypercomplex numbers. As far as I know most functions do not have a hypercomplex extension. So computing gamma of a hypercomplex number does not have an obvious meaning. This is not really a problem now as it deals with it symbolically. But there may require quite a lot of changes in the code which will be a mess to review and coordinate.

**Reply by oscarbenjamin** [reply]
> I think checking if something is hypercomplex is really not that hard

Bear in mind that 99.9% of the time hypercomplex numbers will not be used so this checking is almost always redundant and will complicate every method and slow everything down. This is why I suggested checking the `kind`. The point of the `kind` system is that you don't need *every method* to check for *all* of the kinds of objects that are *not* expected like:
```python
class exp(Function):
    @classmethod
    def eval(cls, arg):
        if is_matrix(arg):
            # do matrix things or error
        elif is_quaternion(arg):
            # do quaternion things or error
        elif ...:
            ...
        else:
             # Presumably we now have an ordinary number
```
Instead you can do
```python
class exp(Function):
    @classmethod
    def eval(cls, arg):
         if arg.kind is NumberKind:
             # Now we know we have an ordinary number
         else:
             # Error out or handle some specific cases
```
The `NumberKind` kind is defined as all the complex numbers, plus the infinities and nan. This is a relatively well-defined scope. Also `NumberKind` is deliberately defined in such a way that it can always be computed efficiently unlike `is_real`, `is_complex` etc. It also represents:

1. All of the kinds of things that could possibly be fully evalf'ed if there are no symbols
2. All of the kinds of things that `solve/solveset` etc can solve for.
3. All of the kinds of things that can be represented as `Poly` for use in `factor` etc
4. All of the kinds of things where the old assumptions predicates make sense

**Reply by JoshZeidner** [reply]
I did take a look at `kind.py`. Seems like you're just reimplementing Multiple Inheritance.

The Quaternion class has `def exp(self):` so it appears that the logic for Exponential function lives inside the Quaternion class.

Really it's something for the future of Sympy, you need a better class structure where the logic and behaviors live inside the classes for the different algebraic term types. I spent a little while examining the code and noticed this.

**Reply by JoshZeidner** [reply]
to @oscargus, Hypercomplex numbers are the future IMHO. They are not a side note to Mathematics. Quaternions have been around longer than most realize. They were shelved in the late 1800s as unusable. In fact Quaternions came before Vectors (they were a simplification). Maxwell originally wrote his equations in Quaternions. There are many fascinating things about Hypercomplex numbers that have yet to be fully accounted for.

I agree that these ideas put more focus on Hypercomplex numbers. They also make the system more linguistically orthogonal. The cost to what already exists in Sympy I can't really tell you. I am considering just taking a swing at this and see how far I get. Appreciate all the help so far.

**Reply by JoshZeidner** [reply]
For instance, the next number system in the C-D construction are [Sedenions](https://en.wikipedia.org/wiki/Sedenion). They are non-alternative. In addition to that in some cases A * B = 0 where A, B are Sedenions and they are not 0 (so-called Zero Divisors). As such there are mysterious holes in the field of Sedenions. The mysterious devolution of algebraic properties continues with each doubling.

My point being, if Sympy could support this it would really be the first of it's kind to give a central role to the C-D construction. This would elevate it's appeal and certainly would attract some interesting characters.

**Reply by oscarbenjamin** [reply]
> I did take a look at `kind.py`. Seems like you're just reimplementing Multiple Inheritance.

The point of `kind` is that it tells you what kind of object an expression represents semantically. In SymPy the type (in programming language terms) is always determined by the head of the expression (the outer-most operation). The `kind` is used to distinguish expressions that have the same head but represent semantically distinct kinds of mathematical object e.g.:
```python
In [12]: I1 = Integral(t, (t, 0, 1))

In [13]: I2 = Integral(Matrix([[t, t], [t**2, t]]), (t, 0, 1))

In [14]: I1
Out[14]: 
1     
⌠     
⎮ t dt
⌡     
0     

In [15]: I2
Out[15]: 
1           
⌠           
⎮ ⎡t   t⎤   
⎮ ⎢     ⎥   
⎮ ⎢ 2   ⎥ dt
⎮ ⎣t   t⎦   
⌡           
0           

In [16]: type(I1)
Out[16]: sympy.integrals.integrals.Integral

In [17]: type(I2)
Out[17]: sympy.integrals.integrals.Integral

In [18]: I1.kind
Out[18]: NumberKind

In [19]: I2.kind
Out[19]: MatrixKind(NumberKind)
```
There is no inheritance at all with kinds and certainly no multiple inheritance. Every `kind` is mutually exclusive with every other so that there is never a need to check for several different properties to determine if an operation can succeed.

**Reply by JoshZeidner** [reply]
yes, but can't this be implemented with multiple inheritance?

I1 has TWO classes Integral and Number
I2 has TWO classes Integral and Matrix (perhaps Matrix is a subclass of Number)

Python is renowned for it's power and flexibility in this regard.

I really don't know much about the Kind system but drawing from experience here seems like you're reimplementing Multiple Inheritance.

> never a need to check for several different properties to determine if an operation can succeed. 

yes I noticed Sympy makes heavy use of these properties. Again Python's OO features can do this quite automatically, also can streamline debugging and such. Python is really not my core language btw.

Thanks for all your help so far!

**Reply by oscarbenjamin** [reply]
> Python is renowned for it's power and flexibility in this regard.

I don't think that you understand how this works in Python. Both `I1` and `I2` are instances of `Integral`. The superclass of `Integral` are determined by the class statement e.g.:
```
class Integral(Expr):
    pass
```
If `Integral` subclasses `Matrix` and `Number` as well then all instances (`I1`, `I2`, etc will be instances of those superclasses as well.

> seems like you're reimplementing Multiple Inheritance.

Firstly I don't think that you understand how multiple inheritance works in Python. Secondly this is explicitly not multiple inheritance. All kinds are mutually exclusive. There isn't even any concept of a subclass with kinds: every object has exactly one kind. It isn't possible for an object to be simultaneously of any two distinct kinds.

**Comment by JoshZeidner** [comment]
@oscarbenjamin, does Sympy handle [Holomorphic Functions](https://en.wikipedia.org/wiki/Holomorphic_function)? couldn't find anything in google search.

**Reply by oscarbenjamin** [reply]
I don't understand the question. What does it mean to "handle" holomorphic functions?

**Reply by JoshZeidner** [reply]
Happy New Years.

That would be functions that take in a complex variable and return a complex variable. They can also return several values for one input as the image shows. The image is the function f(z) = z^(1/4). 

![image](https://user-images.githubusercontent.com/3150100/147844296-2158385e-d8b0-4034-9a9b-ce9c1b630573.png)

a number can have 4 4th roots in the complex field.

![image](https://user-images.githubusercontent.com/3150100/147844252-5c9feb3a-a18f-4779-86fb-d915264a1ea3.png)

**Reply by JoshZeidner** [reply]
I noted before that Sympy does not return multiple complex roots, just one:

https://gist.github.com/JoshZeidner/fe043b055851462b7a5c7de2f5cd21d4

perhaps there is a usage of evalf() that I don't know about?

[Mathematica does show multiple roots.](https://www.wolframalpha.com/input/?i=i%5E%281%2F4%29)

Sympy needs at the minimum support for [Multi-Valued Functions](https://en.wikipedia.org/wiki/Multivalued_function)
